File: 3.0/golang/15.go
package main

import "fmt"

type stack struct {
	arr [][]int
	ln  int
}

func (stack *stack) append(x ...[]int) {
	for _, v := range x {
		stack.arr = append(stack.arr, v)
		stack.ln += 1
	}
}

func (stack *stack) pop() []int {
	num := stack.arr[stack.ln-1]
	stack.arr = stack.arr[:stack.ln-1]
	stack.ln -= 1

	return num
}

func main() {
	var slice [][]int
	stack := stack{arr: slice, ln: 0}

	var N int
	fmt.Scanln(&N)
	ans := make([]int, N)
	for i := range ans {
		ans[i] = -1
	}

	var x int
	for i := 0; i < N; i++ {
		fmt.Scan(&x)

		if stack.ln > 0 {
			for j := stack.ln - 1; j >= 0; j-- {
				if x < stack.arr[j][1] {
					tmp := stack.pop()
					ans[tmp[0]] = i
				} else {
					break
				}
			}
			stack.append([]int{i, x})
		} else {
			stack.append([]int{i, x})
		}
	}

	for _, v := range ans {
		fmt.Print(v, " ")
	}
}

File: 3.0/python/15.py
N = int(input())
pr = list(map(int, input().split()))
stack = []
ans = [-1 for i in range(N)]
ln = 0

for i, v in enumerate(pr):
    if ln > 0:
        for j in reversed(stack):
            if v < j[1]:
                x = stack.pop()
                ln -= 1
                ans[x[0]] = i
            else:
                break

        stack.append([i, v])
        ln += 1
    else:
        stack.append([i, v])
        ln += 1

print(*ans)